Add Success and Failure result classes.

tmertens opened the pull request:
@taylorzr This adds Success and Failure result classes which make Verbalize easier to work with when testing interactors in consumers.  There is no longer any need to know anything about the internals of the Result class to produce a failed result; rather, in test you can now just create `Verbalize::Success` or `Verbalize::Failure` instances.

Also, I added `Failure#error` and deprecated the generic `Result#value` method, which is unnatural to reference when working with a failed result.

 - Deprecate Result#value and Failure#value in favor of Success#value and Failure#error
 - Refactor verbalize_spec to eliminate constant redefined warnings

taylorzr commented on spec/integration/verbalize_spec.rb:
```diff
@@ -64,8 +64,7 @@ def some_method_name
 
-          def call
-          end
+          def call; end
```
Eww, this is a rubocop default?

tmertens replied:
yep, rubocop 0.46

tmertens commented on lib/verbalize/failure.rb:
```diff
@@ -0,0 +1,15 @@
+      super(outcome: :error, value: error)
+    end
+
+    def error
```
We can defer this for now, I will remove until it becomes more apparent as an actual need one way or the other but I haven't loved using `#value` either for failures.  In almost all cases with `Interactor`, people assign failure messages to an `error` or `errors` key on the context object.

tmertens commented:
@taylorzr I reverted the `error` addition and value deprecation.  I also added documentation around testing Verbalize interactors.  See the TODO I added to the stubbing `call!` responses.  It's still a little messy, we should figure out how to make testing a better experience.

tmertens commented:
Also, not sure what you want to do about the TODO.  rubocop is complaining but I don't really want to address it in this PR.

taylorzr commented on README.md:
```diff
@@ -150,6 +150,140 @@ Comparison:
+  input :a
+  
+  def call
+    fail!('${a} is not less than 100!') if a >= 100
```
Should `#{a}` not `${a}` right?

tmertens replied:
oops

taylorzr replied:
Still needs fixed. And actually, this is inconsistent with the previous example using `a`. Don't really care which as long as it's consistent.

taylorzr commented on README.md:
```diff
@@ -150,6 +150,140 @@ Comparison:
+    end
+    
+    it 'returns an error when MyInteractor fails' do
+      failed_result = Verbalize::Failure.new('Y U NO!')
```
Looks like this needs to be removed.

taylorzr replied:
^ right?

taylorzr replied:
Nevermind me, I don't read good.

taylorzr commented on README.md:
```diff
@@ -150,6 +150,140 @@ Comparison:
+      
+      expect {
+        described_class.something
+      }.to raise_error UncaughtThrowError
```
Yea, we'll have to figure out a better way to test this case. Something like `allow(MyInteractor).to receive(:call).and_fail('the failure message, if any')`

taylorzr commented on spec/integration/verbalize_spec.rb:
```diff
@@ -1,13 +1,37 @@
+  end
+
   describe '.verbalize' do
+    it 'returns an ErrorResult on failure' do
```
ErrorResult -> Failure

taylorzr commented on spec/integration/verbalize_spec.rb:
```diff
@@ -1,13 +1,37 @@
+      expect(some_class.call).to be_an_instance_of(Verbalize::Failure)
+    end
+
+    it 'returns a SuccessResult on success' do
```
SuccessResult -> Success

taylorzr commented on lib/verbalize/result.rb:
```diff
@@ -18,7 +18,7 @@ def failed?
 
     def to_ary
-      [outcome, value]
+      [outcome, @value]
```
Don't need this change do we?

tmertens replied:
Since `value` will be going away from the base class as an accessor in v2.0, this needs to access the `@var` directly.

taylorzr commented on spec/integration/verbalize_spec.rb:
```diff
@@ -1,13 +1,37 @@
+    Class.new do
+      include Verbalize
+    end
+  end
```
I would say this base class is unecessary but not a blocker

taylorzr commented:
These coveralls messages are annnnnnnnnoying

taylorzr commented on lib/verbalize/failure.rb:
```diff
@@ -0,0 +1,22 @@
+  class Failure < Result
+    extend Gem::Deprecate
+
+    def initialize(error)
```
This should be named *failure* for consistency right?
